### PR TITLE
vo_kitty: Introduce modern sixel alternative

### LINKED ISSUE #9605: Kitty backend
The kitty terminal emulator has a native graphics protocol supporting image preview.
It does so through one of its default kittens (= extensions) : icat
Here is the relevant documentation for icat: https://sw.kovidgoyal.net/kitty/kittens/icat/

In an issue from 2020, the author of kitty expressed that he doesn't want to work on a video player since videos include sound as well. See here: https://github.com/kovidgoyal/kitty/issues/2947

What I am requesting is that someone takes a quick look at kittys graphics protocol and checks how easy this feature would be to implement an maintain, and if it is as trivial as it appears to me from a users perspective, implement it.
Alternatively, if you find a way to make video work indirectly by piping the vo=image driver through some stuff, that would be a cool placeholder.

Either way, thank you for your great work and have a good christmas season.

PS: I will crossreference this issue on the above mentioned issue regarding video playback in kitty.

### Expected behavior of the wanted feature
--vo=kitty

### Alternative behavior of the wanted feature
--vo=image | ... | icat

### Log file
mpv 0.32.0 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
built on UNKNOWN
ffmpeg library versions:
libavutil       56.51.100
libavcodec      58.91.100
libavformat     58.45.100
libswscale      5.7.100
libavfilter     7.85.100
libswresample   3.7.100
ffmpeg version: 4.3.3-0+deb11u1

### PR DESCRIPTION
**Pull request by mia-0**
See https://sw.kovidgoyal.net/kitty/graphics-protocol/

This makes no attempt at querying terminal features or handling terminal errors, as it would require mpv to pass the response codes from the terminal to the vo instead of interpreting them as keystrokes made by the user and acting very unpredictably.

Tested with kitty and konsole.

Fixes #9605

### CODE REVIEW

File: DOCS/man/vo.rst [hunk 1]
```diff
@@ -383,6 +383,45 @@ Available video output drivers are:
     ``--vo-tct-256=<yes|no>`` (default: no)
         Use 256 colors - for terminals which don't support true color.
 
+``kitty``
+    Graphical output for the terminal, using the kitty graphics protocol.
+    Tested with ``kitty`` and ``konsole``.
```
Comment (sfan5) [inline DOCS/man/vo.rst:388]:
```suggestion
    Tested with kitty and Konsole.
```

File: video/out/vo_kitty.c [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        p->buffer = mmap(NULL, p->buffer_size,
+                         PROT_READ | PROT_WRITE, MAP_SHARED, p->shm_fd, 0);
+
+        if (p->buffer == MAP_FAILED) {
+            MP_WARN(vo, "Failed to mmap shared memory object");
+            shm_unlink(p->shm_path);
+            close(p->shm_fd);
+            return;
+        }
+    }
```
Comment (sfan5) [inline video/out/vo_kitty.c:274]: this has potential for extracting to a function

File: video/out/vo_kitty.c [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+                break;
+        }
+    }
+
+    write_str(cmd);
+    talloc_free(cmd);
+
+#if HAVE_POSIX
+    if (p->opts.use_shm)
+        close_shm(p);
```
Comment (sfan5) [inline video/out/vo_kitty.c:321]: is the shm really meant to be closed and reopened every frame?

Comment (mia-0) [reply video/out/vo_kitty.c:321]: According to the spec, the terminal emulator is supposed to unlink and close it after reading the data.

Comment (sfan5) [reply video/out/vo_kitty.c:328]: Huh.

File: video/out/vo_kitty.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+    }
+
+    free_bufs(vo);
+}
+
+#define OPT_BASE_STRUCT struct priv
+
+const struct vo_driver video_out_kitty = {
+    .name = "kitty",
+    .description = "Video output using the kitty terminal graphics protocol",
```
Comment (sfan5) [inline video/out/vo_kitty.c:405]: one more: this reads kinda redundant otherwise
```suggestion
    .description = "Kitty terminal graphics protocol",
```

**Comment by sfan5** [comment]
Works :+1:

**Pull request closed without merging**